Tasks/SonarQubePostTest/CodeAnalysisFilePathComputation.ps1

```diff
@@ -0,0 +1,170 @@
+$ProjectGuidAndFilePathMap = @{}
+$ComponentKeyAndPathMap = @{}
+$ComponentKeyAndRelativePathCache = @{}
+
+function ConstructComponentKeyAndPathMap($json)
+{
+    foreach ($component in $json.Components)
+    {
+        #Write-Host "Component Key: $($component.key), Path:$($component.path)"
+        $ComponentKeyAndPathMap.Add($component.key, $component.path)
+    }
+}
+
+#returns a path relative to the repo root for a file which has new code analysis issue(s)
+function GetRelativeFilePath($component)
+{
+    if ($ComponentKeyAndRelativePathCache.ContainsKey($component))
+    {
+        $relativeFilePath = $ComponentKeyAndRelativePathCache[$component]
+        Write-Verbose -Verbose "GetRelativeFilePath: Found cached entry, returning data from cache, relativePath:$relativeFilePath"
+
+        return $relativeFilePath
+    }
+
+    #MSBuild runner creates the component value as '[SonarQube project key]:[SonarQube project value]:[MSBuild project guid]:[file name relative to MSBuild project file path]'
+    $tokens = $component.ToString().Split(":")
+
+    if ($tokens -eq $null)
+    {
+        Write-Verbose -Verbose "GetRelativeFilePath: tokens is null, component:$component"
+        return $null
+    }
+    if ($tokens.Count -lt 2) 
+    {
+        Write-Verbose -Verbose "GetRelativeFilePath: tokens.count is less than 2, component:$component"
+        return $null
+    }
+
+    #fair to assume second last token will be guid?
+    $guidToken = $tokens[$tokens.Count - 2]
+    Write-Verbose -Verbose "GetRelativeFilePath: guidToken:$guidToken"
```

Comment:
The code in lines 33 and 40 won't work correctly if there are 3, 5 or more tokens.
We're expecting the token to consist of exactly four sections; if not, then the key is not in the expected form and it would be better to exit.
It might be cleaner to separate out the key parsing and validation logic into a separate function.